web/src/peer_data.ts

```diff
@@ -1,15 +1,77 @@
+import {z} from "zod";
+
 import * as blueslip from "./blueslip.ts";
+import * as channel from "./channel.ts";
 import {LazySet} from "./lazy_set.ts";
+import {page_params} from "./page_params.ts";
 import type {User} from "./people.ts";
 import * as people from "./people.ts";
 import * as sub_store from "./sub_store.ts";
 
 // This maps a stream_id to a LazySet of user_ids who are subscribed.
 const stream_subscribers = new Map<number, LazySet>();
 const fetched_stream_ids = new Set<number>();
+export function has_full_subscriber_data(stream_id: number): boolean {
+    return fetched_stream_ids.has(stream_id);
+}
+const pending_subscriber_requests = new Map<
+    number,
+    {
+        subscribers_promise: Promise<LazySet>;
+        pending_peer_events: {
+            type: "peer_add" | "peer_remove";
+            user_ids: number[];
+        }[];
+    }
+>();
 
 export function clear_for_testing(): void {
     stream_subscribers.clear();
+    fetched_stream_ids.clear();
+}
+
+const fetch_stream_subscribers_response_schema = z.object({
+    subscribers: z.array(z.number()),
+});
+
+export async function maybe_fetch_stream_subscribers(stream_id: number): Promise<LazySet> {
+    if (pending_subscriber_requests.has(stream_id)) {
+        return pending_subscriber_requests.get(stream_id)!.subscribers_promise;
+    }
+    const subscribers_promise = (async () => {
+        let subscribers: number[];
+        try {
+            const xhr = await channel.get({
+                url: `/json/streams/${stream_id}/members`,
+            });
+            subscribers = fetch_stream_subscribers_response_schema.parse(xhr).subscribers;
+        } catch {
+            blueslip.error("Failure fetching channel subscribers", {
+                stream_id,
+            });
+            pending_subscriber_requests.delete(stream_id);
+            // Fall back to what we already have.
+            return get_user_set(stream_id);
+        }
+
+        set_subscribers(stream_id, subscribers);
+        const pending_peer_events = pending_subscriber_requests.get(stream_id)!.pending_peer_events;
+        pending_subscriber_requests.delete(stream_id);
+        for (const event of pending_peer_events) {
+            if (event.type === "peer_add") {
+                bulk_add_subscribers({stream_ids: [stream_id], user_ids: event.user_ids});
+            } else {
+                bulk_remove_subscribers({stream_ids: [stream_id], user_ids: event.user_ids});
+            }
+        }
+        return get_user_set(stream_id);
+    })();
+
+    pending_subscriber_requests.set(stream_id, {
+        subscribers_promise,
+        pending_peer_events: [],
+    });
+    return subscribers_promise;
 }
 
 function get_user_set(stream_id: number): LazySet {
```

Comment:
I think we probably want a naming scheme that makes clear this may be incomplete data.

Maybe call this `get_loaded_subscriber_subset` and `get_full_user_set` => `get_full_subscriber_set`?

Reply:
Great idea, made that change.